prototypes/alpaka_binding/simple_cuda/README.md

```diff
@@ -0,0 +1,41 @@
+# About
+
+The prototype uses pybind11 to create a Python 3 binding from CUDA C++ code. The goal of the implementation is to share memory between the Python and CUDA C++ code and to implement an algorithm in CUDA.
+
+This is the second step in providing a Python binding of an algorithm written with the alpaka library.
+
+# Requirements
+
+* pybind11
+* numpy
+* cmake > 3.14
+* cuda > 9.0
+
+Available via pip or conda.
+
+```bash
+  conda install -c conda-forge pybind11
+  conda install -c anaconda numpy
+  conda install -c anaconda cmake
+  # if CUDA is not provided throw the base environment
+  conda install -c conda-forge cudatoolkit-dev
+```
+
+
+# Install
+
+```bash
+  mkdir build
+  cd build
+  # maybe you need to set: -DPYTHON_EXECUTABLE:FILEPATH=
+  cmake ..
+  make
```

Comment:
Same error as the Alpaka example:

```
(libertem) C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build>msbuild ALL_BUILD.vcxproj
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 09/04/2020 14:16:15.
Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\ALL_BUILD.vcxproj" on node 1 (default targets).
Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\ALL_BUILD.vcxproj" (1) is building "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cu
da\build\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "x64\Debug\ZERO_CHECK\".
  Creating directory "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\".
InitializeBuildStatus:
  Creating "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Checking Build System
FinalizeBuildStatus:
  Deleting file "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\ZERO_CHECK.vcxproj" (default targets).

Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\ALL_BUILD.vcxproj" (1) is building "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cu
da\build\cuBinding.vcxproj" (3) on node 1 (default targets).
PrepareForBuild:
  Creating directory "cuBinding.dir\Debug\".
  Creating directory "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\Debug\".
  Creating directory "cuBinding.dir\Debug\cuBinding.tlog\".
InitializeBuildStatus:
  Creating "cuBinding.dir\Debug\cuBinding.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule C:/Users/weber/Documents/LiberTEM/ptychography/prototypes/alpaka_binding/simple_cuda/CMakeLists.txt
AddCudaCompileDeps:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\cl.exe /E /nologo /showIncludes /TP /D__CUDACC__ /D_WINDOWS /DCMAKE_INTDIR="Debug" /DcuBinding_EXPORTS /DCMAKE_I
  NTDIR="Debug" /DcuBinding_EXPORTS /D_WINDLL /D_MBCS /IC:\Users\weber\.conda\envs\libertem\include /I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin" /I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA
  \v10.2\include" /I. /FIcuda_runtime.h /c C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\binding.cu
Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\cuBinding.vcxproj" (3) is building "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cu
da\build\cuBinding.vcxproj" (3:2) on node 1 (CudaBuildCore target(s)).
CudaBuildCore:
  Compiling CUDA source file ..\binding.cu...
  cmd.exe /C "C:\Users\weber\AppData\Local\Temp\tmp1bc593ced963420b9917ca00e8c21281.cmd"
  "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,compute_30\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tool
  s\MSVC\14.16.27023\bin\HostX86\x64" -x cu  -IC:\Users\weber\.conda\envs\libertem\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Debug -maxrregcount=0  --machine 64 --c
  ompile -cudart static -Xcompiler="/EHsc -Zi -Ob0" -g   -D_WINDOWS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D_WINDLL -D_MBCS -Xcompiler "/EHsc /W3 /nologo /Od /Fdc
  uBinding.dir\Debug\vc141.pdb /FS /Zi /RTC1 /MDd /GR" -o cuBinding.dir\Debug\binding.obj "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\binding.cu"

  (libertem) C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build>"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,comp
  ute_30\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64" -x cu  -IC:\Users\weber\.conda\envs\libertem\include -I"C:\Program Files\NVIDI
  A GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Debug -maxrregcount=0  --machine 64 --compile -cudart static -Xcompiler="/EHsc -Zi -Ob0" -g   -D_WINDOWS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -
  D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D_WINDLL -D_MBCS -Xcompiler "/EHsc /W3 /nologo /Od /FdcuBinding.dir\Debug\vc141.pdb /FS /Zi /RTC1 /MDd /GR" -o cuBinding.dir\Debug\binding.obj "C:\Users\weber\Documents\
  LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\binding.cu"
c:\users\weber\.conda\envs\libertem\include\pybind11\cast.h(1495): error : expression must be a pointer to a complete object type [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\bu
ild\cuBinding.vcxproj]

  1 error detected in the compilation of "C:/Users/weber/AppData/Local/Temp/tmpxft_0000242c_00000000-5_binding.cpp4.ii".
  binding.cu
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\BuildCustomizations\CUDA 10.2.targets(764,9): error MSB3721: The command ""C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2
\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,compute_30\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64" -x cu  -IC:\Users\weber
\.conda\envs\libertem\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Debug -maxrregcount=0  --machine 64 --compile -cudart static -Xcompiler="/EHsc -Zi -Ob0" -g   -D_WIN
DOWS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D_WINDLL -D_MBCS -Xcompiler "/EHsc /W3 /nologo /Od /FdcuBinding.dir\Debug\vc141.pdb /FS /Zi /RTC1 /MDd /GR" -o cuBindi
ng.dir\Debug\binding.obj "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\binding.cu"" exited with code 2. [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\
simple_cuda\build\cuBinding.vcxproj]
Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\cuBinding.vcxproj" (CudaBuildCore target(s)) -- FAILED.

Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\cuBinding.vcxproj" (default targets) -- FAILED.

Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.


Build FAILED.

"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\cuBinding.vcxproj" (default target) (3) ->
"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\build\cuBinding.vcxproj" (CudaBuildCore target) (3:2) ->
(CudaBuildCore target) ->
  c:\users\weber\.conda\envs\libertem\include\pybind11\cast.h(1495): error : expression must be a pointer to a complete object type [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\
build\cuBinding.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\BuildCustomizations\CUDA 10.2.targets(764,9): error MSB3721: The command ""C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10
.2\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,compute_30\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64" -x cu  -IC:\Users\web
er\.conda\envs\libertem\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Debug -maxrregcount=0  --machine 64 --compile -cudart static -Xcompiler="/EHsc -Zi -Ob0" -g   -D_W
INDOWS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D"CMAKE_INTDIR=\"Debug\"" -DcuBinding_EXPORTS -D_WINDLL -D_MBCS -Xcompiler "/EHsc /W3 /nologo /Od /FdcuBinding.dir\Debug\vc141.pdb /FS /Zi /RTC1 /MDd /GR" -o cuBin
ding.dir\Debug\binding.obj "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\simple_cuda\binding.cu"" exited with code 2. [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_bindin
g\simple_cuda\build\cuBinding.vcxproj]

    0 Warning(s)
    2 Error(s)

Time Elapsed 00:00:13.74
```

Reply:
We found a solution for the problem. It's a bug in pybind11 described here: https://github.com/pytorch/pytorch/issues/11004
I will open a issue at pybind11 to fix it.
After that a second error appears. The cuBinding.pyd is not build. Renaming cuBinding.dll to cuBinding.pyd solve the problem. I know the problem and will fix it in the CMakeLists.txt

But finally it works :-)